File: Problem2.py
#Time complexity : O(n)
#Space complexity: O(1)
#Works on leetcode : yes
#Approach : Since either A or B needs to be all same, we just need to check for A[0] or B[0] as c in all further ith positions. 
#If at A[i] or B[i], we dont have A[0] or B[0] - we return -1. Otherwise we count how many times either A[0] or B[0] appear
#in top or bottom and use these count to get minimum number of flips. 

class Solution(object):
    def minDominoRotations(self, A, B):
        """
        :type A: List[int]
        :type B: List[int]
        :rtype: int
        """
        if len(A)!=len(B):
            return -1
        for c in {A[0],B[0]}:
            top,bottom = 0,0
            for i in range(len(A)):
                if c not in {A[i],B[i]}:
                    break
                top +=1 if c!=B[i] else 0
                bottom +=1 if c!=A[i] else 0
            else:
                return min(top,bottom)
        return -1

File: Problem_2.java
// Time complexity - O(n)
// Space complexity - O(1)

class Solution {
    public int minDominoRotations(int[] A, int[] B) {
        if(A == null || A.length == 0 || B == null || B.length == 0) return 0;
        int maxFreq = 0;
        for(int i = 0 ; i < 7; i++){
            int count = 0;
            for(int j = 0 ; j < A.length; j++){
                if(A[j] == i) count++;
                if(B[j] == i) count++;
                if(count >= A.length){
                    maxFreq = i;
                    break;
                }
            }            
        }
        
        int aRot = 0; int bRot = 0;
        for(int i = 0 ; i < A.length; i++){
            if(A[i] != maxFreq &&  B[i] != maxFreq) return -1;
            else if(A[i] != maxFreq) aRot++;
            else if(B[i] != maxFreq) bRot++;
        }
        return Math.min(aRot,bRot);
    }
}
